src/system.rs

```diff
@@ -1,7 +1,105 @@
 use std::marker::PhantomData;
+use std::ops::Deref;
 
 use {ResourceId, Resources};
 
+/// A trait for accessing read/write multiple resources from a system. This can be used
+/// to create dynamic systems that don't specify what they fetch at compile-time.
+///
+/// For compile-time system data this will all be done for you using `StaticAccessor`.
+pub trait Accessor: Sized {
+    /// Tries to create a new instance of this type. This one returns `Some` in case there is a
+    /// default, otherwise the system needs to override `System::accessor`.
+    fn try_new() -> Option<Self>;
+
+    /// A list of [`ResourceId`]s the bundle
+    /// needs read access to in order to
+    /// build the target resource bundle.
+    ///
+    /// # Contract
+    ///
+    /// Exactly return the dependencies you're going to `fetch`! Doing otherwise *will* cause a
+    /// panic.
+    ///
+    /// This method is only executed once,
+    /// thus the returned value may never change
+    /// (otherwise it has no effect).
+    ///
+    /// [`ResourceId`]: struct.ResourceId.html
+    fn reads(&self) -> Vec<ResourceId>;
+
+    /// A list of [`ResourceId`]s the bundle
+    /// needs write access to in order to
+    /// build the target resource bundle.
+    ///
+    /// # Contract
+    ///
+    /// Exactly return the dependencies you're going to `fetch`! Doing otherwise *will* cause a
+    /// panic.
+    ///
+    /// This method is only executed once,
+    /// thus the returned value may never change
+    /// (otherwise it has no effect).
+    ///
+    /// [`ResourceId`]: struct.ResourceId.html
+    fn writes(&self) -> Vec<ResourceId>;
+}
+
+impl Accessor for () {
+    fn try_new() -> Option<Self> {
+        None
+    }
+    fn reads(&self) -> Vec<ResourceId> {
+        Vec::new()
+    }
+    fn writes(&self) -> Vec<ResourceId> {
+        Vec::new()
+    }
+}
+
+impl<T: ?Sized> Accessor for PhantomData<T> {
+    fn try_new() -> Option<Self> {
+        None
+    }
+    fn reads(&self) -> Vec<ResourceId> {
+        Vec::new()
+    }
+    fn writes(&self) -> Vec<ResourceId> {
+        Vec::new()
+    }
+}
+
+/// Either an `Accessor` of the system `T` or a reference to it.
+pub enum AccessorCow<'a, 'b, T>
```

Comment:
Why does this need it's own `Cow`? Why not use stds?

Reply:
`std`s requires it to be `Clone`

Reply:
It's not 100% a copy on write type, because it just means it's either a reference or an owned instance.

Reply:
`RefOrOwned`, `OwnedOrNot`, `RefOrNot`, `MaybeOwned`, `Borrowned`.
Cannot think good name for it.